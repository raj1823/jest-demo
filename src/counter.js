import React from 'react';
import {Text, TouchableOpacity} from 'react-native';
import {useSelector, useDispatch} from 'react-redux';

const Counter = () => {
  const dispatch = useDispatch();
  const count = useSelector(state => state.count);

  return (
    <>
      <Text testID={'text'}>You pressed the button {count} times</Text>
      <TouchableOpacity
        testID={'button'}
        onPress={() => dispatch({type: 'INCREMENT'})}>
        +1
      </TouchableOpacity>
    </>
  );
};

export default Counter;
