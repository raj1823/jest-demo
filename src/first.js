import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  Dimensions,
  Image,
  Animated,
  PanResponder,
  TouchableOpacity,
} from 'react-native';
import {connect} from 'react-redux';
import axios from 'axios'
import {increase, decrease} from '../reducers/actions';
const SCREEN_HEIGHT = Dimensions.get('window').height;
const SCREEN_WIDTH = Dimensions.get('window').width;

export class First extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      counter: 1,
    };
  }

  componentDidMount() {
    this.updateCounter();
  }
  updateCounter = () => {
    this.setState({counter: this.state.counter + 1});
  };

  dummyApi = async callback => {
    const res = await axios.get('https://jsonplaceholder.typicode.com/todos/1')
    if (res) callback && callback(res.data);
    else callback && callback(res)
  };

  render() {
    return (
      <View
        style={{
          flex: 1,
          alignItems: 'center',
          //justifyContent: 'center',
          backgroundColor: '#fff',
        }}>
        <View style={{alignItems: 'center'}}>
          <Text style={{fontSize: 30, marginVertical: 30}}>
            {`Counter: ${this.props.counter}`}
          </Text>
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
            }}>
            <TouchableOpacity
              onPress={() => {
                this.props.increase();
              }}
              style={{
                backgroundColor: 'green',
                height: 40,
                width: 80,
                justifyContent: 'center',
                alignItems: 'center',
                margin: 15,
              }}>
              <Text style={{fontSize: 15, color: '#fff'}}>{'Increase'}</Text>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => {
                this.props.decrease();
              }}
              style={{
                backgroundColor: 'green',
                height: 40,
                width: 80,
                justifyContent: 'center',
                alignItems: 'center',
                margin: 15,
              }}>
              <Text style={{fontSize: 15, color: '#fff'}}>{'Decrease'}</Text>
            </TouchableOpacity>
          </View>
        </View>

        <TouchableOpacity
          onPress={() => this.props.navigation.navigate('second')}
          style={{
            height: 70,
            width: 120,
            backgroundColor: 'blue',
            justifyContent: 'center',
            alignItems: 'center',
          }}>
          <Text style={{fontSize: 20, color: '#fff'}}>{'Next'}</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const mapStateToProps = state => ({
  counter: state.CounterReducer.counter,
});
const mapDispatchToProps = dispatch => ({
  increase: () => increase(dispatch),
  decrease: () => decrease(dispatch),
});
export default connect(mapStateToProps, mapDispatchToProps)(First);
