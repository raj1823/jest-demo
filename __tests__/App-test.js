import 'react-native';
import React from 'react';
import {First} from '../src/first';
import {Second} from '../src/second';
import Counter from '../src/counter';
import {Provider} from 'react-redux';
import {shallow, configure} from 'enzyme';
import thunk from 'redux-thunk';
import Adapter from 'enzyme-adapter-react-16';

import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {render, cleanup, fireEvent} from '@testing-library/react-native';
configure({adapter: new Adapter()});
afterEach(() => {
  cleanup();
});
jest.setTimeout(10000);

// Note: test renderer must be required after react-native.
// JEST DOCUMENTATION : https://jestjs.io/docs/api

//USE CASES :
// Shallow: create and test component only (no children)
// Mount: same as shallow but mounts with children and parent/host component, allows lifecycle methods
// Render: outputs the html given by the component, including children

describe('<First/>', () => {
  const mockStore = configureStore([]);

  it('snapshot test', () => {
    const tree = render(<First />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('component Did mount', () => {
    const wrapper = shallow(<First />).instance();

    wrapper.state.counter = 1;
    wrapper.componentDidMount();

    expect(wrapper.state.counter).toBe(2);
  });

  it('mocking class functions', () => {
    const wrapper = shallow(<First />).instance();
    const updateCounterMock = jest.spyOn(wrapper, 'updateCounter');
    wrapper.componentDidMount();
    expect(updateCounterMock).toHaveBeenCalledTimes(1);
  });

  it('mocking props', () => {
    const props = {
      increase: function () {},
    };
    const wrapper = shallow(<Second increase={jest.fn()} />).instance();
    wrapper.increaseCounter();
    expect(wrapper.props.increase).toHaveBeenCalledTimes(1);
  });

  it('mocking function with returning value', () => {
    const wrapper = shallow(<Second />).instance();
    wrapper.updateMyValue = jest.fn().mockReturnValue(10);
    expect(wrapper.updateMyValue()).toBe(10);
  });

  it('mocking function with calling value', () => {
    const wrapper = shallow(<Second />).instance();
    const spyFunction = jest.spyOn(wrapper, 'callWithJestFunction');
    wrapper.callWithJestFunction(20);
    expect(spyFunction).toHaveBeenCalledWith(20);
  });

  // Asynchronous calls testing

  it('Asynchronous callbacks testing', done => {
    const wrapper = shallow(<First />).instance();
    wrapper.dummyApi(callback);
    function callback(data) {
      try {
        expect(data).toStrictEqual({
          userId: 1,
          id: 1,
          title: 'delectus aut autem',
          completed: false,
        });
        done();
      } catch (error) {
        done(error);
      }
    }
  });

  // timer functions mocking (setTimeout, setIntervals etc.)
  it('timer mock', () => {
    jest.useFakeTimers();
    const wrapper = shallow(<First />).instance();
    jest.spyOn(global, 'setTimeout');
    wrapper.timerGame();
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000);
  });

  it('should display current count', () => {
    const store = mockStore({count: 5});
    const rendered = render(
      <Provider store={store}>
        <Counter />
      </Provider>,
    );
    const textComponent = rendered.getByTestId('text');

    expect(textComponent.props.children).toContain(5);
  });

  it('should dispatch increment action', () => {
    const store = mockStore({count: 5});
    const rendered = render(
      <Provider store={store}>
        <Counter />
      </Provider>,
    );
    const buttonComponent = rendered.getByTestId('button');

    fireEvent(buttonComponent, 'press');
    const actions = store.getActions();
    expect(actions.length).toBe(1);
    expect(actions[0].type).toEqual('INCREMENT');
  });
});
