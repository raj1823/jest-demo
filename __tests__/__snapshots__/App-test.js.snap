// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<First/> snapshot test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 30,
          "marginVertical": 30,
        }
      }
    >
      Counter: undefined
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "height": 40,
            "justifyContent": "center",
            "margin": 15,
            "opacity": 1,
            "width": 80,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 15,
            }
          }
        >
          Increase
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "height": 40,
            "justifyContent": "center",
            "margin": 15,
            "opacity": 1,
            "width": 80,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 15,
            }
          }
        >
          Decrease
        </Text>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "blue",
        "height": 70,
        "justifyContent": "center",
        "opacity": 1,
        "width": 120,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
        }
      }
    >
      Next
    </Text>
  </View>
</View>
`;
